File: web/src/screens/levelScreen.ts
import { Button } from '../constants.js';
import { rgb, selectFeedback } from '../utils.js';
import { Screen } from '../screen.js';
import { Input } from '../input.js';
import { GameState } from '../gameState.js';
import { Resource } from '../resource.js';
import { Music } from '../media/music.js';
import type { Animation } from '../media/animation.js';
import type { BitmapFont } from '../media/font.js';
import { LevelDescription } from '../levelDescription.js';
import { Room } from '../room.js';
import { loadRoom } from '../roomLoader.js';
import { LevelSelectScreen } from './levelSelectScreen.js';

export class LevelScreen extends Screen {
  private readonly darken: Animation;
  private isExit = false;
  private readonly exitBg: Animation;
  private readonly font: BitmapFont;
  private readonly hand: Animation;
  private readonly levelDesc: LevelDescription;
  private readonly room: Room;
  private selected = 0;

  constructor(desc: LevelDescription) {
    super();
    this.levelDesc = desc;
    this.darken = Resource.getAnimation('data/images/darken.bmp');
    this.exitBg = Resource.getAnimation('data/images/level_exit_background.bmp', 1);
    this.font = Resource.getFont('data/images/font.bmp');
    this.hand = Resource.getAnimation('data/images/hand.bmp', 1);

    Music.playSong(desc.musicFile);
    this.room = loadRoom(desc.levelFile);
  }

  onDraw(ctx: CanvasRenderingContext2D): void {
    ctx.fillStyle = rgb(57, 56, 41);
    ctx.fillRect(0, 0, 320, 240);

    this.room.onDraw(ctx);

    const coins = GameState.getInt('coins');
    const coinsStr = '[x' + coins;

    ctx.fillStyle = rgb(57, 56, 41);
    ctx.fillRect(0, 0, 320, 10);
    this.font.draw(ctx, coinsStr, 1, 1);

    if (this.isExit) {
      for (let y = 0; y < 15; y++)
        for (let x = 0; x < 20; x++)
          this.darken.drawFrame(ctx, 0, x * 16, y * 16);

      this.exitBg.drawFrame(ctx, 0,
        160 - this.exitBg.getFrameWidth() / 2,
        120 - this.exitBg.getFrameHeight() / 2);
      this.font.draw(ctx, 'CONTINUE', 130, 110);
      this.font.draw(ctx, 'EXIT LEVEL', 130, 120);
      this.hand.drawFrame(ctx, 0, 120, 110 + this.selected * 10);
    }
  }

  onLogic(): void {
    if (this.room.isCompleted()) {
      if (this.levelDesc.levelFile === 'data/rooms/olof.txt')
        LevelSelectScreen.bossLevelCompleted = true;
      GameState.put(this.levelDesc.levelFile, 1);
      this.exit();
    }

    if (!this.isExit) this.room.onLogic();

    if (Input.isPressed(Button.EXIT) && !this.isExit) {
      selectFeedback();
      this.isExit = true;
    } else if (Input.isPressed(Button.EXIT) && this.isExit) {
      selectFeedback();
      this.isExit = false;
    }

    if (!this.isExit) return;

    if (Input.isPressed(Button.DOWN)) {
      this.selected++;
      if (this.selected > 1) this.selected = 1;
      else selectFeedback();
    }
    if (Input.isPressed(Button.UP)) {
      this.selected--;
      if (this.selected < 0) this.selected = 0;
      else selectFeedback();
    }
    if (Input.isPressed(Button.FIRE)) {
      selectFeedback();
      if (this.selected === 0) this.isExit = false;
      else if (this.selected === 1) this.exit();
    }
  }
}

File: wii/src/screens/level_screen.cpp
#include "screens/level_screen.h"
#include "resource.h"
#include "media/animation.h"
#include "media/font.h"
#include "media/sound.h"
#include "media/music.h"
#include "input.h"
#include "game_state.h"
#include "constants.h"
#include "room.h"
#include "room_loader.h"
#include "screens/level_select_screen.h"
#include <string>

LevelScreen::LevelScreen(const LevelDescription& desc)
    : levelDesc_(desc)
{
    darken_ = Resource::getAnimation("data/images/darken.bmp");
    exitBg_ = Resource::getAnimation("data/images/level_exit_background.bmp", 1);
    font_ = Resource::getFont("data/images/font.bmp");
    hand_ = Resource::getAnimation("data/images/hand.bmp", 1);

    Music::playSong(desc.musicFile);
    room_ = loadRoom(desc.levelFile);
}

LevelScreen::~LevelScreen() {
    delete room_;
}

void LevelScreen::onDraw(SDL_Renderer* renderer) {
    SDL_SetRenderDrawColor(renderer, 57, 56, 41, 255);
    SDL_Rect bg = {0, 0, SCREEN_WIDTH, SCREEN_HEIGHT};
    SDL_RenderFillRect(renderer, &bg);

    room_->onDraw(renderer);

    int coins = GameState::getInt("coins");
    std::string coinsStr = "[x" + std::to_string(coins);

    SDL_SetRenderDrawColor(renderer, 57, 56, 41, 255);
    SDL_Rect topBar = {0, 0, SCREEN_WIDTH, 10};
    SDL_RenderFillRect(renderer, &topBar);
    font_->draw(renderer, coinsStr, 1, 1);

    if (isExit_) {
        for (int y = 0; y < 15; y++)
            for (int x = 0; x < 20; x++)
                darken_->drawFrame(renderer, 0, x * 16, y * 16);

        exitBg_->drawFrame(renderer, 0,
            160 - exitBg_->getFrameWidth() / 2,
            120 - exitBg_->getFrameHeight() / 2);
        font_->draw(renderer, "CONTINUE", 130, 110);
        font_->draw(renderer, "EXIT LEVEL", 130, 120);
        hand_->drawFrame(renderer, 0, 120, 110 + selected_ * 10);
    }
}

void LevelScreen::onLogic() {
    if (room_->isCompleted()) {
        if (levelDesc_.levelFile == "data/rooms/olof.txt")
            LevelSelectScreen::bossLevelCompleted = true;
        GameState::put(levelDesc_.levelFile, 1);
        exit();
    }

    if (!isExit_) room_->onLogic();

    if (Input::isPressed(Button::EXIT) && !isExit_) {
        Sound::playSample("data/sounds/select");
        isExit_ = true;
    } else if (Input::isPressed(Button::EXIT) && isExit_) {
        Sound::playSample("data/sounds/select");
        isExit_ = false;
    }

    if (!isExit_) return;

    if (Input::isPressed(Button::DOWN)) {
        selected_++;
        if (selected_ > 1) selected_ = 1;
        else Sound::playSample("data/sounds/select");
    }
    if (Input::isPressed(Button::UP)) {
        selected_--;
        if (selected_ < 0) selected_ = 0;
        else Sound::playSample("data/sounds/select");
    }
    if (Input::isPressed(Button::FIRE)) {
        Sound::playSample("data/sounds/select");
        if (selected_ == 0) isExit_ = false;
        else if (selected_ == 1) exit();
    }
}
